pkg/logs/input/kubernetes/launcher.go

```diff
@@ -0,0 +1,219 @@
+		return
+	}
+	l.addSourcesFromPod(pod)
+}
+
+// removeSources removes a new log-source from a service
+func (l *Launcher) removeSources(service *service.Service) {
+	containerID := service.GetEntityID()
+	if source, exists := l.sourcesByContainer[containerID]; exists {
+		delete(l.sourcesByContainer, containerID)
+		l.sources.RemoveSource(source)
+	}
+}
+
+// addSourcesFromPod creates new log-sources for each container of the pod.
+// it checks if the sources already exist to avoid tailing twice the same
+// container when pods have multiple containers
+func (l *Launcher) addSourcesFromPod(pod *kubelet.Pod) {
+	for _, container := range pod.Status.Containers {
+		containerID := container.ID
+		if _, exists := l.sourcesByContainer[containerID]; exists {
+			continue
+		}
+		source, err := l.getSource(pod, container)
+		if err != nil {
+			log.Warnf("Invalid configuration for pod %v, container %v: %v", pod.Metadata.Name, container.Name, err)
+			continue
+		}
+		l.sourcesByContainer[containerID] = source
+		l.sources.AddSource(source)
+	}
+}
+
+// kubernetesIntegration represents the name of the integration.
+const kubernetesIntegration = "kubernetes"
+
+// getSource returns a new source for the container in pod.
+func (l *Launcher) getSource(pod *kubelet.Pod, container kubelet.ContainerStatus) (*config.LogSource, error) {
+	var cfg *config.LogsConfig
+	if annotation := l.getAnnotation(pod, container); annotation != "" {
+		configs, err := config.ParseJSON([]byte(annotation))
+		if err != nil || len(configs) == 0 {
+			return nil, fmt.Errorf("could not parse kubernetes annotation %v", annotation)
+		}
+		cfg = configs[0]
+	} else {
+		cfg = &config.LogsConfig{
+			Source:  kubernetesIntegration,
+			Service: kubernetesIntegration,
+		}
+	}
+	cfg.Type = config.FileType
+	cfg.Path = l.getPath(pod, container)
+	cfg.Identifier = container.ID
+	cfg.Tags = append(cfg.Tags, l.getTags(container)...)
+	if err := cfg.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid kubernetes annotation: %v", err)
+	}
+	if err := cfg.Compile(); err != nil {
+		return nil, fmt.Errorf("could not compile kubernetes annotation: %v", err)
+	}
+	return config.NewLogSource(l.getSourceName(pod, container), cfg), nil
+}
+
+// configPath refers to the configuration that can be passed over a pod annotation,
+// this feature is commonly named 'ad' or 'autodicovery'.
+// The pod annotation must respect the format: ad.datadoghq.com/<container_name>.logs: '[{...}]'.
+const (
+	configPathPrefix = "ad.datadoghq.com"
+	configPathSuffix = "logs"
+)
+
+// getConfigPath returns the path of the logs-config annotation for container.
+func (l *Launcher) getConfigPath(container kubelet.ContainerStatus) string {
+	return fmt.Sprintf("%s/%s.%s", configPathPrefix, container.Name, configPathSuffix)
+}
+
+// getAnnotation returns the logs-config annotation for container if present.
+func (l *Launcher) getAnnotation(pod *kubelet.Pod, container kubelet.ContainerStatus) string {
```

Comment:
Okay for keeping the annotation here for now but I think we should add a FIXME and create a card to remove this logic as it's done in `autodiscovery`.